# LAB | Hyperparameter Tuning

**Load the data**

Finally step in order to maximize the performance on your Spaceship Titanic model.

The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

So far we've been training and evaluating models with default values for hyperparameters.

Today we will perform the same feature engineering as before, and then compare the best working models you got so far, but now fine tuning it's hyperparameters.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [3]:
# drop null values
df_space = spaceship.copy()
df_space = df_space.dropna()

# transforming Cabin values
df_space["Cabin"] = df_space["Cabin"].str[0]

# Drop PassengerId and Name
df_space = df_space.drop(columns = ["PassengerId", "Name", "Cabin"], axis = 1)

#For non-numerical columns, do dummies.
df_space_final = pd.get_dummies(df_space, drop_first = True)

In [4]:
df_space_final

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0     39.0          0.0        0.0           0.0     0.0     0.0        False   
1     24.0        109.0        9.0          25.0   549.0    44.0         True   
2     58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3     33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4     16.0        303.0       70.0         151.0   565.0     2.0         True   
...    ...          ...        ...           ...     ...     ...          ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0        False   
8689  18.0          0.0        0.0           0.0     0.0     0.0        False   
8690  26.0          0.0        0.0        1872.0     1.0     0.0         True   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0        False   
8692  44.0        126.0     4688.0           0.0     0.0    12.0         True   

      HomePlanet_Europa  HomePlanet_Mars  CryoSleep_True  \
0                  True            False           False   
1                 False            False           False   
2                  True            False           False   
3                  True            False           False   
4                 False            False           False   
...                 ...              ...             ...   
8688               True            False           False   
8689              False            False            True   
8690              False            False           False   
8691               True            False           False   
8692               True            False           False   

      Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_True  
0                         False                     True     False  
1                         False                     True     False  
2                         False                     True      True  
3                         False                     True     False  
4                         False                     True     False  
...                         ...                      ...       ...  
8688                      False                    False      True  
8689                       True                    False     False  
8690                      False                     True     False  
8691                      False                    False     False  
8692                      False                     True     False  

[6606 rows x 13 columns]

- Now let's use the best model we got so far in order to see how it can improve when we fine tune it's hyperparameters.

In [5]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, root_mean_squared_error

In [6]:
X = df_space_final.drop(['Transported'], axis=1)
y = df_space_final['Transported'].astype(int)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state=0)

In [7]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [8]:
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(n_estimators = 100,
                             max_depth = 20)

forest.fit(X_train_scaled, y_train)
forest_score = forest.score(X_test_scaled, y_test)
print("Random Forest accuracy:", forest_score)

Random Forest accuracy: 0.7806354009077155


- Evaluate your model

In [9]:
from sklearn.metrics import classification_report

y_pred = forest.predict(X_test_scaled)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.78      0.78      0.78       661
           1       0.78      0.78      0.78       661

    accuracy                           0.78      1322
   macro avg       0.78      0.78      0.78      1322
weighted avg       0.78      0.78      0.78      1322



**Grid/Random Search**

For this lab we will use Grid Search.

- Define hyperparameters to fine tune.

In [10]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}

- Run Grid Search

In [11]:
gs = GridSearchCV(forest, param_grid=param_grid, cv=4, n_jobs=-1, verbose=2)

gs.fit(X_train_scaled, y_train)

print("Best Parameters:", gs.best_params_)
print("Grid Search Score:", gs.best_score_)

Fitting 4 folds for each of 108 candidates, totalling 432 fits
Best Parameters: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 2, 'n_estimators': 100}
Grid Search Score: 0.8054504163512491


- Evaluate your model

The model improved from 0.78 to 0.81
And the Best Parameters are: 
- 'max_depth': 10
- 'min_samples_leaf': 4
- 'min_samples_split': 2
- 'n_estimators': 100